source/postcard-c/postcard.h

```diff
@@ -0,0 +1,1184 @@
+        i++;
+
+        *value |= ((uint64_t)(byte & 0x7F)) << shift;
+        if (!(byte & 0x80))
+            return POSTCARD_SUCCESS;
+
+        shift += 7;
+        if (shift > 63)
+            return POSTCARD_ERROR_OVERFLOW;
+    }
+
+    return POSTCARD_ERROR_OVERFLOW;
+}
+
+// Decode a signed varint (zigzag encoded)
+inline static postcard_error_t decode_signed_varint(postcard_slice_t* slice,
+    int64_t* value,
+    size_t max_bytes)
+{
+    uint64_t zigzag;
+    postcard_error_t err = decode_unsigned_varint(slice, &zigzag, max_bytes);
+    if (err != POSTCARD_SUCCESS)
+        return err;
+
+    // Zigzag decoding: (n >> 1) ^ (-(n & 1))
+    *value = (zigzag >> 1) ^ (-(zigzag & 1));
+    return POSTCARD_SUCCESS;
+}
+
+inline postcard_error_t postcard_encode_bool(postcard_slice_t* slice,
+    bool value)
+{
+    if (!slice || !slice->data)
+        return POSTCARD_ERROR_INVALID_INPUT;
+    if (slice->len >= slice->capacity)
+        return POSTCARD_ERROR_BUFFER_TOO_SMALL;
+
+    slice->data[slice->len++] = value ? 0x01 : 0x00;
+    return POSTCARD_SUCCESS;
+}
+
+inline postcard_error_t postcard_encode_u8(postcard_slice_t* slice,
+    uint8_t value)
+{
+    if (!slice || !slice->data)
+        return POSTCARD_ERROR_INVALID_INPUT;
+    if (slice->len >= slice->capacity)
+        return POSTCARD_ERROR_BUFFER_TOO_SMALL;
+
+    slice->data[slice->len++] = value;
+    return POSTCARD_SUCCESS;
+}
+
+inline postcard_error_t postcard_encode_i8(postcard_slice_t* slice,
+    int8_t value)
+{
+    if (!slice || !slice->data)
+        return POSTCARD_ERROR_INVALID_INPUT;
+    if (slice->len >= slice->capacity)
+        return POSTCARD_ERROR_BUFFER_TOO_SMALL;
+
+    slice->data[slice->len++] = (uint8_t)value;
+    return POSTCARD_SUCCESS;
+}
+
+inline postcard_error_t postcard_encode_u16(postcard_slice_t* slice,
+    uint16_t value)
+{
+    return encode_unsigned_varint(slice, value, 3);
+}
+
+inline postcard_error_t postcard_encode_i16(postcard_slice_t* slice,
+    int16_t value)
+{
+    return encode_signed_varint(slice, value, 3);
+}
+
+inline postcard_error_t postcard_encode_u32(postcard_slice_t* slice,
+    uint32_t value)
+{
+    return encode_unsigned_varint(slice, value, 5);
+}
+
+inline postcard_error_t postcard_encode_i32(postcard_slice_t* slice,
+    int32_t value)
+{
+    return encode_signed_varint(slice, value, 5);
+}
+
+inline postcard_error_t postcard_encode_u64(postcard_slice_t* slice,
+    uint64_t value)
+{
+    return encode_unsigned_varint(slice, value, 10);
+}
+
+inline postcard_error_t postcard_encode_i64(postcard_slice_t* slice,
+    int64_t value)
+{
+    return encode_signed_varint(slice, value, 10);
+}
+
+inline postcard_error_t postcard_encode_f32(postcard_slice_t* slice,
+    float value)
+{
+    if (!slice || !slice->data)
+        return POSTCARD_ERROR_INVALID_INPUT;
+    if (slice->len + 4 > slice->capacity)
+        return POSTCARD_ERROR_BUFFER_TOO_SMALL;
+
+    uint32_t bits;
+    memcpy(&bits, &value, sizeof(bits));
```

Comment:
Should we use the memcpy in the standard library or make the caller pick the memcpy?

One pattern I often see is that a library has a seperate memcpy function, like a `postcard_memcpy`. This would let the caller define it's own memcpy implementation.

Reply:
I've been going back and forth on what the best behavior is here. Personally I like to use a minimal libc (i.e newlib) on embedded devices, but I can understand not wanting to rely on that. I think the ergonomic benefits for non-embedded development are important though. 

A good way to do it might be to use `__has_include` to check if `string` is available and use that version of memcpy if it exists. Alternatively we could use a preprocessor definition to allow the user to optionally override the memcpy impl